federation/src/main/scala/caliban/federation/EntityResolver.scala

```diff
@@ -20,10 +21,42 @@ object EntityResolver {
     resolver: A => ZQuery[R, CalibanError, Option[T]]
   )(implicit schema: Schema[R, T]): EntityResolver[R] =
     new EntityResolver[R] {
-      override def resolve(value: InputValue): ZQuery[R, CalibanError, Step[R]] =
-        ZQuery.fromEither(implicitly[ArgBuilder[A]].build(value)).flatMap { arg =>
-          resolver(arg).map(_.fold[Step[R]](Step.NullStep)(schema.resolve))
-        }
+      override def resolve(value: InputValue): Step[R] =
+        QueryStep(
+          ZQuery.fromEither(implicitly[ArgBuilder[A]].build(value)).flatMap { arg =>
+            resolver(arg).map(_.fold[Step[R]](Step.NullStep)(schema.resolve))
+          }
+        )
```

Comment:
Avoids the extra flatmap and `ZQuery.fromEither`:

```suggestion
        ArgBuilder[A].build(value) match {
          case Right(arg)  =>
            val q = resolver(arg).map {
              case Some(value) => schema.resolve(value)
              case _           => Step.NullStep
            }
            Step.QueryStep(q)
          case Left(error) => Step.FailureStep(error)
        }
```